### PR TITLE
feat(topolotree): perf improvements and better deploy logic

### PR DESCRIPTION
**Pull request by shadaj**
feat(topolotree): perf improvements and better deploy logic

### CODE REVIEW

**Comment by shadaj** [comment]
DO NOT MERGE

File: hydro_cli/src/core/hydroflow_crate/mod.rs [hunk 1]
```diff
@@ -38,7 +38,8 @@ pub struct HydroflowCrate {
     /// Configuration for the ports that this service will listen on a port for.
     port_to_bind: HashMap<String, ServerStrategy>,
 
-    built_binary: Option<JoinHandle<Result<BuiltCrate>>>,
+    building_binary: Option<JoinHandle<Result<BuiltCrate>>>,
+    built_binary: Option<BuiltCrate>,
```
Comment (MingweiSamuel) [inline hydro_cli/src/core/hydroflow_crate/mod.rs:42]: maybe using something like `MaybeDone` would be cleaner with these two fields
https://docs.rs/futures/latest/futures/future/fn.maybe_done.html

Comment (MingweiSamuel) [reply hydro_cli/src/core/hydroflow_crate/mod.rs:42]: Or async once cell maybe https://docs.rs/async-once-cell/latest/async_once_cell/struct.OnceCell.html

**Review by MingweiSamuel — APPROVED** [review]
Seems like this undoes some of the previous commit? :P

**Comment by shadaj** [comment]
> Seems like this undoes some of the previous commit? :P

Yeah, had to abandon some of the fault tolerance logic for now 😢

**Pull request merged**